Add communication with react native webviews post (#113)
closeio/making.close.com: _posts/2024-01-15-communicating-with-react-native-webviews/index.md

---
title: 'Communicating with React Native Web Views'
date: 2024-01-15
permalink: /posts/react-native-webviews
author: Trey Cucco
thumbnail: ''
metaDescription: ''
tags: [Engineering, React, Mobile App]
---

In 2023 we started work on our mobile app. The first thing we had to do was pick
a platform: native app vs. progressive web app? platform-specific vs.
cross-platform? If we picked cross-platform: React Native? Flutter? Something

> else?

Author: That's a really excellent point Scott, I'll make some changes in this direction!

> else?

Author: @essmahr I've rewritten the intro. I'd appreciate if you'd read through that and give me some feedback on it!

> else?

Reviewer: Yeah man looks great! No further feedback

The tl;dr of that post is: we picked React Native to build a cross-platform app,
and we decided to use a web view to include most of the functionality from our
web app, and then augment that with native navigation and functionality.

There are several options for using web views in React Native, and we picked the
excellent
[react-native-webview](https://github.com/react-native-webview/react-native-webview)
library. It's a mature, actively developed, and popular library.

Since we wanted to do more than _just_ have a web view we knew we'd need two-way
communication between the native app and the web app. Fortunately, React Native
WebView provides several primitives for communicating between the native app and
the web app.

In this post I'm going to talk through the communication primitives provided by
React Native WebView, some of the potential
[developer footguns](https://en.wiktionary.org/wiki/footgun) we noticed, and the
patterns we developed to make sure our native app and web app could evolve
independently without breaking.

## Communication Patterns

React Native WebView provides several ways of communicating between native and

> web. If you want a full description the project provides a

Reviewer 2: ```suggestion
web. If you want a full description, the project provides a
```

[dedicated section of the docs](https://github.com/react-native-webview/react-native-webview/blob/master/docs/Guide.md#communicating-between-js-and-native)
for that. We're going to focus on two of the options: `injectJavaScript` to
communicate to web from native, and the `postMessage` / `onMessage` pair, which
allows communication to native from web.

### Receiving Communications in Native

To send information from the web app to the native app, you first attach a
handler to the
[`onMessage prop`](https://github.com/react-native-webview/react-native-webview/blob/master/docs/Reference.md#onmessage)

> of the web view. This handler receives an event with a string data prop. The web

Reviewer 2: ```suggestion
of the WebView. This handler receives an event with a string data prop. The web
```

little bit of a nit, but might as well call it by its class name?

> of the web view. This handler receives an event with a string data prop. The web
> view fires this event whenever the web app calls a global function provided by
> the web view: `window.ReactNativeWebView.postMessage`.

Reviewer 2: ```suggestion
of the `WebView`. This handler receives an event with a string data prop. The `WebView` fires this event whenever the web app calls a global function provided by
the `WebView`: `window.ReactNativeWebView.postMessage`.
```

> of the web view. This handler receives an event with a string data prop. The web
> view fires this event whenever the web app calls a global function provided by
> the web view: `window.ReactNativeWebView.postMessage`.
> 
> So in React Native you'd create a web view component like this:

Reviewer 2: little bit of a nit, but might as well call it by its class name?

```suggestion
of the `WebView`. This handler receives an event with a string data prop. The `WebView` fires this event whenever the web app calls a global function provided by
the `WebView`: `window.ReactNativeWebView.postMessage`.

So in React Native you'd create a web view component like this:
```

```jsx
<WebView
  onMessage={(event) => {
    console.log(event.data);
  }}
  {...otherProps}
/>
```

And in your web app JavaScript you can do something like this:

```TypeScript
const sendMessageToNative = (message: string) => {
	window.ReactNativeWebView.postMessage(message);
}

sendMessageToNative("Hello, React Native!");
```

Then, back in React Native, you'd see "Hello, React Native!" logged out whenever
the web app was run.

### Receiving Communications in Web

Things are not as straight forward when communicating to the web app from native
app. Rather than exposing an event-based system, React Native WebView allows you

> to inject arbitrary JavaScript code into the web view at any time. This means

Reviewer 2: ```suggestion
to inject arbitrary JavaScript code into the `WebView` at any time. This means
```

that will take the action and payload, put them in the correct format, and
serialize it all into valid JavaScript code to be injected:

```TypeScript
const buildMessageJavaScript = (action: string, payload: any) => {
	const message = JSON.stringify({ action, payload });
	// Stringify the message a second time to escape quotes etc.
	const safeString = JSON.stringify(message);

	return `window.onMessageFromRN(${safeString});`;
};

const postMessageToWebApp = (
	webViewRef: MutableRefObject<WebView>,
	action: string,
	payload: any,
) => {
	webViewRef.current.injectJavaScript(
		buildMessageJavaScript(action, payload),
	);
}
```

With that in place, we now had the basis for a two-way message bridge to send
structured messages between the native app and the web app.

### Strongly Typed Messages

We used this pattern to work on our app for a few weeks, but the lack of type
safety on our messages became a real pain: it was easy to misspell an action
name, or to give the wrong data in the payload. So we decided to strongly type
our messages and make sure that on both sides, whenever we send a message, the
action string belonged to the pre-identified set of actions, and the payload

> attached to it conformed to the expected shape for that action.

Reviewer: maybe tighter: `So, we decided to to strongly type the messaging system. Now, whenever we send a message (from web or native) the action string must belong to a pre-identified set of actions, and the payload must conform to the expected shape for that action.`

```TypeScript
// Defines the actions and payloads that the web app can send to the native app
export type FromWebActions = {
	callInitiated: {
		leadId: string;
		contactId: string;
		phoneNumber: string;
	};
	organizationChanged: {
		organizationId: string;
	};
	appReady: {
		organizationId: string;
	}
	// many more actions
};

export type FromWebActionName = keyof FromWebActions;

// Defines the actions and payloads that the native app can send to the web app
export type FromNativeActions = {
	profilePopoverToggled: null;
	routeChanged: {
		pathname: string;
		options: { replace?: boolean; };
	};
	searchHidden: null;
	// many more actions
};

export type FromNativeActionName = keyof FromNativeActions;
```

And we updated our message handling methods like this:

```TypeScript
export const postMessageToRN = <T extends FromWebActionName>(
	action: T,
	payload: FromWebActions[T],
) => {
	// same implementation as above
}

export const registerRNHandler = <T extends FromNativeActionName>(
	action: T,
	callback: (payload: FromNativeActions[T]) => void,
) => {
	// same implementation as above
}
```

With that (and something similar on the native side) we now had type-safe

> message passing between the two systems! If we tried to give an action name that

Reviewer: present tense? `If we try to give ... Typescript warns us`

> wasn't recognized, or if we gave a payload with the wrong shape or data,

Reviewer 2: Out of curiosity, and maybe it's out of scope for this blog post, but what is the impact in the case where the app hasn't been updated and one side of the application doesn't know of the type of message that is being passed to it?

TypeScript would warn us.

This _greatly_ improved the DX of sending messages between the two systems, and
caught silly bugs quickly.

### Sharing the type definitions

We set up the type definitions so they could be used as-is between the two
systems, but we still needed to figure out how to share those type definitions.

> We thought of several ways we could share them, but in the ended we decided to
> just manually synchronize the file between the two code bases.

Reviewer: ```suggestion
Ultimately, we decided to just manually synchronize the file between the two codebases.
```

> This works very well because when we're making changes to it on one side, we're
> almost always making changes on the other side as well, adding new handlers or
> adding new message posters.

Reviewer: ```suggestion
Since making changes to it on one side almost always involves making changes on the other side (e.g. adding new handlers, or adding new message posters) this is adequate for now.
```

## Limiting Knowledge Between Systems

We want to keep the two systems as decoupled as possible. Being able to
independently deploy the web app and native app was a hard requirement for us.

> The message passing system helps greatly with this, because the web app doesn't
> have to care _how_ the native app will do something (like initiate a phone call)

Reviewer: ```suggestion
The advantage of a message-based communication API is that the web app doesn't
have to care _how_ the native app will do something (like initiate a phone call)
```

it just has to tell it "the user pressed a button to start a call" and the
native app will receive that message and Do The Right Thing.

This even means that the implementation can change on one side and the other app
doesn't have to be updated or re-deployed.

However, there's still one more problem we ran in to that wasn't solved by the
message passing system alone: when we add new features to the native app, and it
expects new messages from the web app, how do we coordinate that? Let's taking
calling as an example.

> When we implemented native calling, we needed the web app to send a message to

Reviewer: this section might deserve it's own header?
